### PR TITLE
Scala blog post

### PR DESCRIPTION
**Pull request by iceddante**
I had some issues learning scala, particularly in that I didn't know what I was getting into when I started that journey. This blog post represents a syllabus for learning that I think would help other OO developers that want to get started there. 

I'd like to hear feedback from Ippon Scala devs on what they think and how it can be improved.

### CODE REVIEW

**Comment by fsarradin** [comment]
Hi Randeep,

I really appreciate your post, especially the way you suggest direction to follow to learn Scala, instead of giving closed answer.

As a senior scala dev, I propose to my coworkers exercises, that help to explore some concepts and discuss about them. That's the topic of my [last post](http://blog.ippon.fr/2018/06/07/de-linteret-de-proposer-des-exercices-de-code-dans-vos-equipes-exemple-en-scala/) (in French btw). I also do some live coding. My coworkers use to appreciate it and their skill improve very quickly by this way.

For exercises and courses, here are some links:
- https://www.scala-exercises.org/
- http://aperiodic.net/phil/scala/s-99/
- https://en.coursera.org/learn/progfun1

If you want to explore FP more deeply, I suggest:
- https://alvinalexander.com/scala/functional-programming-simplified-book (book)
- https://www.manning.com/books/functional-programming-in-scala (book)
- https://gist.github.com/jdegoes/97459c0045f373f4eaf126998d8f65dc (a Gist made by John De Goes; use this resource to see where you are in your learning path at first ^^)
- listening to some talks about FP and Scala on youtube, reading blog on this topic, and look around on twitter for relevant guys to follow
- learning Haskell

**Comment by iceddante** [comment]
Thanks you for the feedback. These are some excellent resources. I also forgot to link Odersky's fp Scala course on coursera which I'll link to as well. Yes, on further consideration the blog gives all emphasis on reading and study and not on programming so I'll make sure to emphasize that as well

File: posts/scala-for-the-arrogant.md [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+[**Scala Best Practices: Pure Functions**](https://dzone.com/articles/scala-best-practices-pure-functions-1) at DZone.com
+
+#### Tips for the Arrogant developer
+*A side effect of worrying about side-effects can be sweaty palms and headaches. Scala tries
+to mitigate these symptoms with built-in operators that help manage state when things go wrong
+so don't get discouraged and start looking at how you can streamline your own functions and methods
+to bring them closer to purity.*
+ 
+### Algebraic Data Types
+How smart will you appear at your next job interview if your start talking about
```
Comment (dsharpe) [inline posts/scala-for-the-arrogant.md]: your -> you

File: posts/scala-for-the-arrogant.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+ 
+# Course Materials
+## Essential Scala
+This online PDF provides an excellent overview of the Scala language and its features. It does not necessarily emphasize the functional model of doing things but gives you all the tools you need to be a great Scala developer in all disciplines.
+
+## Functional and Reactive Domain Modeling
+I highly recommend this book. It's a Scala book disguised as a bible on Functional programming and domain modeling. The result is you can learn a lot of best practices in these fields while learning how Scala correlates to them.
+
+## Programming in Scala
+Written by the creator of Scala, Martin Odersky, this text is a comprehensive guide to Scala. I don't think it is organized as effectively for Java developers fro the old and new schools, but you would not be amiss to have it on your shelf.
```
Comment (dsharpe) [inline posts/scala-for-the-arrogant.md]: fro -> for

File: posts/scala-for-the-arrogant.md [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+## Functional programming
+Functional programming has been described many different ways. I have always
+considered it to be the creation and composition of **rules** to describe a
+system.  In this context, rules are functions:
+*`f(x) = y`*
+for example, tells us that when we supply input x into the function we get y every time. From a mathematics
+perspective that is great. But the astute computer scientist and the practical computer
+engineer will observe that things are seldom so simple. If our function has to fetch
+data from a database, how do we react to deadlocks? Can these rules work in a
+multi-threaded environment with mutable objects? Deciding how to manage the theory of fucntional programming with practical realities is not easy and Scala takes specific steps to address that.
```
Comment (aseigneurin) [inline posts/scala-for-the-arrogant.md]: fucntional -> functional

File: posts/scala-for-the-arrogant.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+### Apply Method
+A very simple concept that has a very commonplace implementation. You should know how it works and how it relates to the [Factory pattern](https://www.journaldev.com/1392/factory-design-pattern-in-java) in Java.
+
+#### See also: 
+*the `unapply` method for deriving an object from a parameter.*
+
+### Mutable variables vs the Copy method
+While Scala tends to encourage immutability with some of its constructs and conventions, developers that are not yet ready to give up the ole setter method can still follow a traditional approach, while the native `copy` method provides a simple construct for creating cloned modified instances of existing objects.
+- How **case classes** come with built-in support for `copy` and how updates can be communicated via copy.
```
Comment (aseigneurin) [inline posts/scala-for-the-arrogant.md]: Need a blank line before this line

File: posts/scala-for-the-arrogant.md [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+list of concepts to study to understand not just how to program in scala, but *why* to
+do so in a specific manner. When approached this way you will learn many new concepts
+and different ways of writing code, managing side effects and more. The arrogant
+developer will probably find themselves being quickly humbled at the breadth of it all,
+a testament to the language and its community of developers.
+ 
+Let's get started!
+ 
+# The path to learning Scala
+1. Adopt the right mindset
```
Comment (aseigneurin) [inline posts/scala-for-the-arrogant.md]: Numbered and un-numbered lists shouldn't have spaces between items

File: posts/scala-for-the-arrogant.md [hunk 6]
```diff
@@ -0,0 +1,231 @@
+---
```
[the remark this answers is not in the document]
Comment (iceddante) [reply posts/scala-for-the-arrogant.md]: I'm using a md editor and it shows up fine there... if you still experience issues with it let me know and maybe you can fix it after the merge.

File: posts/scala-for-the-arrogant.md [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+The answer lies in the intention of the Scala language. As its creator, Martin Odersky, [put it in an interview] (https://www.signifytechnology.com/blog/2016/10/interview-with-scala-creator-martin-odersky):
+> Scala was designed to show that a fusion of
+functional and object-oriented programming is possible and practical." If we
+understand this to be the motivation we can set the proper intention for our
+study of Scala: when a developer learns Scala they are not just learning a new
+language, but an approach to developing algorithms, data modeling, and design
+with the most concise and efficient object-oriented and Functional design.
+
+The arrogant developer might have years of OO design under their belt, they may
```
[the remark this answers is not in the document]
Comment (iceddante) [reply posts/scala-for-the-arrogant.md]: Yes- just more gender neutral :-D

File: posts/scala-for-the-arrogant.md (hunk 5 above)
Comment (iceddante) [reply posts/scala-for-the-arrogant.md]: Not sure I follow here: check out the updated commit and lmk if you think it still has issues.

**Pull request merged**